src/main/java/org/folio/innreach/domain/service/BasicService.java

```diff
@@ -0,0 +1,41 @@
+package org.folio.innreach.domain.service;
+
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import org.springframework.retry.annotation.Backoff;
+import org.springframework.retry.annotation.Retryable;
+
+import org.folio.innreach.domain.exception.ResourceVersionConflictException;
+
+public interface BasicService<K, R> extends UpdateTemplateWithFinder<K, R> {
+
+  R create(R aRecord);
+
+  R update(R aRecord);
+
+  Optional<R> find(K key);
+
+  @Retryable(value = ResourceVersionConflictException.class,
+      maxAttemptsExpression = "#{${retryable-update.on-conflict.retry-attempts}}",
+      backoff = @Backoff(delayExpression = "#{${retryable-update.on-conflict.retry-interval-ms}}"),
+      listeners = {"retryMonitoringListener"})
+  default Optional<R> changeAndUpdate(K key, UpdateOperation<R> change) {
+    return update(key, change.andThen(this::update));
+  }
+
+  @Retryable(value = ResourceVersionConflictException.class,
+      maxAttemptsExpression = "#{${retryable-update.on-conflict.retry-attempts}}",
+      backoff = @Backoff(delayExpression = "#{${retryable-update.on-conflict.retry-interval-ms}}"),
+      listeners = {"retryMonitoringListener"})
+  default Optional<R> changeAndUpdate(K key, Supplier<? extends RuntimeException> notFoundExceptionSupplier,
+      UpdateOperation<R> change) {
+    return update(key, finder().toRequired(notFoundExceptionSupplier), change.andThen(this::update));
+  }
```

Comment:
can't we simply relay on `Optional.orElseThrow` where needed instead of passing notFound handler as an argument?

Reply:
that's what I've done initially, but Sonar detected a bug of not using a return value..
and on the second thought, I also decided that it's better to state during the method call that a result from finder cannot be empty

Reply:
lets then pass it as the last param at least?